File: packages/sdk/src/agent-outbound.ts
import { FLUXY_MAX_MESSAGE_LENGTH } from "./room-rest";

export interface AgentOutboundMessageInput {
  userId: string;
  content: string;
  parentId?: number | null;
  attachments?: unknown[];
}

export interface AgentOutboundValidationResult {
  valid: boolean;
  error?: string;
  content?: string;
  parentId?: number | null;
}

export function validateAgentOutboundMessage(
  input: AgentOutboundMessageInput,
): AgentOutboundValidationResult {
  const userId = input.userId?.trim();
  if (!userId) {
    return { valid: false, error: "userId is required" };
  }
  if (userId.length > 128) {
    return { valid: false, error: "userId exceeds maximum length" };
  }

  if (typeof input.content !== "string") {
    return { valid: false, error: "content must be a string" };
  }
  const trimmed = input.content.trim();
  if (!trimmed) {
    return { valid: false, error: "content cannot be empty" };
  }
  if (trimmed.length > FLUXY_MAX_MESSAGE_LENGTH) {
    return {
      valid: false,
      error: `content exceeds maximum length of ${FLUXY_MAX_MESSAGE_LENGTH} characters`,
    };
  }

  let parentId: number | null = null;
  if (input.parentId != null) {
    if (!Number.isFinite(input.parentId) || input.parentId < 1) {
      return { valid: false, error: "parentId must be a positive message id" };
    }
    parentId = Math.floor(input.parentId);
  }

  const attachments = input.attachments;
  if (attachments != null) {
    if (!Array.isArray(attachments)) {
      return { valid: false, error: "attachments must be an array" };
    }
    if (attachments.length > 20) {
      return { valid: false, error: "attachments cannot exceed 20 items" };
    }
  }

  return { valid: true, content: trimmed, parentId };
}

export function buildAgentOutboundWsPayload(
  input: AgentOutboundMessageInput,
): Record<string, unknown> {
  const validated = validateAgentOutboundMessage(input);
  if (!validated.valid) {
    throw new Error(validated.error ?? "Invalid agent outbound message");
  }
  return {
    type: "message",
    userId: input.userId.trim(),
    content: validated.content,
    parentId: validated.parentId ?? null,
    attachments: Array.isArray(input.attachments) ? input.attachments : [],
  };
}

File: packages/flutter-sdk/lib/src/agent_outbound.dart
const int kFluxyMaxMessageLength = 10000;

class AgentOutboundValidationResult {
  final bool valid;
  final String? error;
  final String? content;
  final int? parentId;
  const AgentOutboundValidationResult({required this.valid, this.error, this.content, this.parentId});
}

AgentOutboundValidationResult validateAgentOutboundMessage({
  required String userId,
  required String content,
  int? parentId,
  List<dynamic>? attachments,
}) {
  final trimmedUserId = userId.trim();
  if (trimmedUserId.isEmpty) return AgentOutboundValidationResult(valid: false, error: 'userId is required');
  if (trimmedUserId.length > 128) return AgentOutboundValidationResult(valid: false, error: 'userId exceeds maximum length');
  if (content.isEmpty) return AgentOutboundValidationResult(valid: false, error: 'content cannot be empty');
  final trimmed = content.trim();
  if (trimmed.isEmpty) return AgentOutboundValidationResult(valid: false, error: 'content cannot be empty');
  if (trimmed.length > kFluxyMaxMessageLength) return AgentOutboundValidationResult(valid: false, error: 'content exceeds maximum length of $kFluxyMaxMessageLength characters');
  int? validParentId;
  if (parentId != null) {
    if (parentId < 1) return AgentOutboundValidationResult(valid: false, error: 'parentId must be a positive message id');
    validParentId = parentId;
  }
  if (attachments != null && attachments.length > 20) return AgentOutboundValidationResult(valid: false, error: 'attachments cannot exceed 20 items');
  return AgentOutboundValidationResult(valid: true, content: trimmed, parentId: validParentId);
}

Map<String, dynamic> buildAgentOutboundWsPayload({
  required String userId,
  required String content,
  int? parentId,
  List<dynamic>? attachments,
}) {
  final validated = validateAgentOutboundMessage(userId: userId, content: content, parentId: parentId, attachments: attachments);
  if (!validated.valid) throw Exception(validated.error ?? 'Invalid agent outbound message');
  return {'type': 'message', 'userId': userId.trim(), 'content': validated.content, 'parentId': validated.parentId, 'attachments': attachments ?? []};
}
